File: apps/server/src/host/acp/mcp_bridge.rs
//! Loopback HTTP MCP bridge injected into ACP `session/new` / `session/load` (t3code parity).

use agent_client_protocol::schema::v1::{HttpHeader, McpServer, McpServerHttp};
use axum::body::Bytes;
use axum::extract::State;
use axum::http::{HeaderMap, StatusCode};
use axum::response::IntoResponse;
use axum::routing::post;
use axum::Router;
use serde_json::{json, Value};
use std::net::SocketAddr;
use std::sync::{Arc, Mutex, OnceLock};
use tokio::sync::oneshot;
use uuid::Uuid;

#[derive(Clone)]
struct BridgeState {
    token: String,
    workspace_root: Arc<Mutex<Option<String>>>,
}

struct BridgeHandle {
    endpoint: String,
    token: String,
    workspace_root: Arc<Mutex<Option<String>>>,
    _shutdown: oneshot::Sender<()>,
}

static BRIDGE: OnceLock<Mutex<Option<BridgeHandle>>> = OnceLock::new();

fn bridge_slot() -> &'static Mutex<Option<BridgeHandle>> {
    BRIDGE.get_or_init(|| Mutex::new(None))
}

/// Ensure the process-wide loopback MCP bridge is running; return ACP `mcpServers` list.
pub fn ensure_mcp_servers(workspace_root: Option<&str>) -> Vec<McpServer> {
    let slot = bridge_slot();
    let mut guard = match slot.lock() {
        Ok(guard) => guard,
        Err(_) => return Vec::new(),
    };
    if guard.is_none() {
        match spawn_bridge() {
            Ok(handle) => *guard = Some(handle),
            Err(error) => {
                tracing::warn!(%error, "failed to start ACP MCP bridge");
                return Vec::new();
            }
        }
    }
    let Some(handle) = guard.as_ref() else {
        return Vec::new();
    };
    if let Some(root) = workspace_root {
        if let Ok(mut cwd) = handle.workspace_root.lock() {
            *cwd = Some(root.to_string());
        }
    }
    let server = McpServer::Http(
        McpServerHttp::new("gharargah", handle.endpoint.clone()).headers(vec![HttpHeader::new(
            "Authorization",
            format!("Bearer {}", handle.token),
        )]),
    );
    vec![server]
}

fn spawn_bridge() -> Result<BridgeHandle, String> {
    let token = Uuid::new_v4().to_string();
    let workspace_root = Arc::new(Mutex::new(None));
    let state = BridgeState {
        token: token.clone(),
        workspace_root: workspace_root.clone(),
    };
    let app = Router::new()
        .route("/", post(handle_mcp))
        .route("/mcp", post(handle_mcp))
        .with_state(state);

    let listener = std::net::TcpListener::bind("127.0.0.1:0").map_err(|e| e.to_string())?;
    listener.set_nonblocking(true).map_err(|e| e.to_string())?;
    let addr = listener.local_addr().map_err(|e| e.to_string())?;
    let endpoint = format!("http://127.0.0.1:{}/mcp", addr.port());
    let (shutdown_tx, shutdown_rx) = oneshot::channel::<()>();

    std::thread::Builder::new()
        .name("gharargah-mcp-bridge".into())
        .spawn(move || {
            let rt = match tokio::runtime::Builder::new_current_thread()
                .enable_all()
                .build()
            {
                Ok(rt) => rt,
                Err(error) => {
                    tracing::error!(%error, "MCP bridge runtime failed");
                    return;
                }
            };
            rt.block_on(async move {
                let listener = match tokio::net::TcpListener::from_std(listener) {
                    Ok(listener) => listener,
                    Err(error) => {
                        tracing::error!(%error, "MCP bridge listener failed");
                        return;
                    }
                };
                let server = axum::serve(
                    listener,
                    app.into_make_service_with_connect_info::<SocketAddr>(),
                );
                tokio::select! {
                    result = server => {
                        if let Err(error) = result {
                            tracing::warn!(%error, "MCP bridge stopped");
                        }
                    }
                    _ = shutdown_rx => {}
                }
            });
        })
        .map_err(|e| e.to_string())?;

    Ok(BridgeHandle {
        endpoint,
        token,
        workspace_root,
        _shutdown: shutdown_tx,
    })
}

async fn handle_mcp(
    State(state): State<BridgeState>,
    headers: HeaderMap,
    body: Bytes,
) -> impl IntoResponse {
    let auth = headers
        .get(axum::http::header::AUTHORIZATION)
        .and_then(|v| v.to_str().ok())
        .unwrap_or("");
    let expected = format!("Bearer {}", state.token);
    if auth != expected {
        return (
            StatusCode::UNAUTHORIZED,
            json!({"jsonrpc":"2.0","error":{"code":-32001,"message":"unauthorized"},"id":null})
                .to_string(),
        );
    }

    let request: Value = match serde_json::from_slice(&body) {
        Ok(value) => value,
        Err(_) => {
            return (
                StatusCode::BAD_REQUEST,
                json!({"jsonrpc":"2.0","error":{"code":-32700,"message":"parse error"},"id":null})
                    .to_string(),
            );
        }
    };

    let id = request.get("id").cloned().unwrap_or(Value::Null);
    let method = request.get("method").and_then(Value::as_str).unwrap_or("");
    let result = match method {
        "initialize" => json!({
            "protocolVersion": "2024-11-05",
            "capabilities": { "tools": {} },
            "serverInfo": { "name": "gharargah", "version": env!("CARGO_PKG_VERSION") }
        }),
        "notifications/initialized" | "initialized" => {
            // Notification — empty ack body is fine for streamable HTTP clients that still POST.
            return (StatusCode::ACCEPTED, String::new());
        }
        "ping" => json!({}),
        "tools/list" => json!({
            "tools": [{
                "name": "gharargah_ping",
                "description": "Health check for the Gharargah host MCP bridge",
                "inputSchema": { "type": "object", "properties": {} }
            }, {
                "name": "gharargah_workspace_root",
                "description": "Return the workspace root bound to the current ACP session",
                "inputSchema": { "type": "object", "properties": {} }
            }]
        }),
        "tools/call" => {
            let name = request
                .pointer("/params/name")
                .and_then(Value::as_str)
                .unwrap_or("");
            match name {
                "gharargah_ping" => json!({
                    "content": [{ "type": "text", "text": "pong" }]
                }),
                "gharargah_workspace_root" => {
                    let root = state
                        .workspace_root
                        .lock()
                        .ok()
                        .and_then(|g| g.clone())
                        .unwrap_or_default();
                    json!({
                        "content": [{ "type": "text", "text": root }]
                    })
                }
                _ => {
                    return (
                        StatusCode::OK,
                        json!({
                            "jsonrpc": "2.0",
                            "id": id,
                            "error": { "code": -32601, "message": format!("unknown tool: {name}") }
                        })
                        .to_string(),
                    );
                }
            }
        }
        _ => {
            return (
                StatusCode::OK,
                json!({
                    "jsonrpc": "2.0",
                    "id": id,
                    "error": { "code": -32601, "message": format!("method not found: {method}") }
                })
                .to_string(),
            );
        }
    };

    (
        StatusCode::OK,
        json!({ "jsonrpc": "2.0", "id": id, "result": result }).to_string(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ensure_mcp_servers_returns_http_entry() {
        let servers = ensure_mcp_servers(Some("/tmp/ws"));
        assert_eq!(servers.len(), 1);
        match &servers[0] {
            McpServer::Http(http) => {
                assert_eq!(http.name, "gharargah");
                assert!(http.url.contains("127.0.0.1"));
                assert!(!http.headers.is_empty());
            }
            _ => panic!("expected HTTP MCP server"),
        }
    }
}

File: apps/agent-server/src/provider/mcp-bridge.ts
/**
 * Loopback HTTP MCP bridge injected into ACP session/new|load (t3code / Rust parity).
 * Tools: gharargah_ping, gharargah_workspace_root.
 */
import http from "node:http"
import { randomUUID } from "node:crypto"

type BridgeHandle = {
  endpoint: string
  token: string
  workspaceRoot: string | null
  server: http.Server
}

let bridge: BridgeHandle | null = null

function jsonRpc(id: unknown, result: unknown): string {
  return JSON.stringify({ jsonrpc: "2.0", id: id ?? null, result })
}

function jsonRpcError(id: unknown, code: number, message: string): string {
  return JSON.stringify({ jsonrpc: "2.0", id: id ?? null, error: { code, message } })
}

function spawnBridge(): BridgeHandle {
  const token = randomUUID()
  const handle: BridgeHandle = {
    endpoint: "",
    token,
    workspaceRoot: null,
    server: null as unknown as http.Server,
  }

  const server = http.createServer((req, res) => {
    const url = req.url ?? "/"
    if (req.method !== "POST" || (url !== "/" && url !== "/mcp" && !url.startsWith("/mcp?"))) {
      res.writeHead(404)
      res.end()
      return
    }
    const chunks: Buffer[] = []
    req.on("data", c => chunks.push(c))
    req.on("end", () => {
      const auth = req.headers.authorization ?? ""
      if (auth !== `Bearer ${token}`) {
        res.writeHead(401, { "content-type": "application/json" })
        res.end(jsonRpcError(null, -32001, "unauthorized"))
        return
      }
      let request: { id?: unknown; method?: string; params?: { name?: string } }
      try {
        request = JSON.parse(Buffer.concat(chunks).toString("utf8")) as typeof request
      } catch {
        res.writeHead(400, { "content-type": "application/json" })
        res.end(jsonRpcError(null, -32700, "parse error"))
        return
      }
      const id = request.id ?? null
      const method = request.method ?? ""
      if (method === "notifications/initialized" || method === "initialized") {
        res.writeHead(202)
        res.end()
        return
      }
      let result: unknown
      switch (method) {
        case "initialize":
          result = {
            protocolVersion: "2024-11-05",
            capabilities: { tools: {} },
            serverInfo: { name: "gharargah", version: "0.0.1" },
          }
          break
        case "ping":
          result = {}
          break
        case "tools/list":
          result = {
            tools: [
              {
                name: "gharargah_ping",
                description: "Health check for the Gharargah host MCP bridge",
                inputSchema: { type: "object", properties: {} },
              },
              {
                name: "gharargah_workspace_root",
                description: "Return the workspace root bound to the current ACP session",
                inputSchema: { type: "object", properties: {} },
              },
            ],
          }
          break
        case "tools/call": {
          const name = request.params?.name ?? ""
          if (name === "gharargah_ping") {
            result = { content: [{ type: "text", text: "pong" }] }
          } else if (name === "gharargah_workspace_root") {
            result = {
              content: [{ type: "text", text: handle.workspaceRoot ?? "" }],
            }
          } else {
            res.writeHead(200, { "content-type": "application/json" })
            res.end(jsonRpcError(id, -32601, `unknown tool: ${name}`))
            return
          }
          break
        }
        default:
          res.writeHead(200, { "content-type": "application/json" })
          res.end(jsonRpcError(id, -32601, `unknown method: ${method}`))
          return
      }
      res.writeHead(200, { "content-type": "application/json" })
      res.end(jsonRpc(id, result))
    })
  })

  server.listen(0, "127.0.0.1")
  const addr = server.address()
  const port = typeof addr === "object" && addr ? addr.port : 0
  handle.endpoint = `http://127.0.0.1:${port}/mcp`
  handle.server = server
  return handle
}

/** ACP `mcpServers` list for session/new|load|resume. */
export function ensureMcpServers(workspaceRoot?: string | null): unknown[] {
  if (process.env.GHARARGAH_AGENT_MOCK === "1") {
    return [
      {
        name: "gharargah-mock",
        command: "/usr/bin/true",
        args: [],
        env: [],
      },
    ]
  }
  if (!bridge) {
    try {
      bridge = spawnBridge()
    } catch (err) {
      console.warn("[mcp-bridge] failed to start:", err)
      return []
    }
  }
  if (workspaceRoot) bridge.workspaceRoot = workspaceRoot
  return [
    {
      type: "http",
      name: "gharargah",
      url: bridge.endpoint,
      headers: [{ name: "Authorization", value: `Bearer ${bridge.token}` }],
    },
  ]
}

export async function closeMcpBridge(): Promise<void> {
  if (!bridge) return
  const b = bridge
  bridge = null
  await new Promise<void>((resolve, reject) => {
    b.server.close(err => (err ? reject(err) : resolve()))
  })
}
